torch/_functorch/_aot_autograd/fx_passes.py

```diff
@@ -0,0 +1,202 @@
+from collections import defaultdict
+from typing import Dict, List
+
+import torch
+from torch import fx
+from torch.utils import _pytree as pytree
+
+
+def is_primal(node: fx.Node) -> bool:
+    return node.op == "placeholder" and "tangents" not in str(node.target)
+
+
+def move_resize_zero_to_end_of_graph(graph: fx.Graph) -> None:
+    node_list = list(graph.nodes)
+    return_node = node_list[-1]
+    assert return_node.target == "output"
+    resize_nodes_to_insert_at_end = []
+    node_to_usage_map = defaultdict(list)
+    for i, n in enumerate(node_list):
+        if n.op == "call_function":
+            if (
+                n.target is torch.ops.inductor.resize_storage_bytes_.default
+                and n.args[1] == 0
+            ):
+                resize_nodes_to_insert_at_end.append((n, i))
+            args_flat = pytree.arg_tree_leaves(*n.args, **n.kwargs)
+            for arg in args_flat:
+                node_to_usage_map[arg].append(i)
+    # Assumptions: the size-0 tensor X can then be used only in the following ops:
+    # 1. `aten.set_(X, ...)`
+    # 2. `aten.set_(..., X)`
+    # 3. `inductor.resize_storage_bytes_(X, 0)`
+    # 4. graph output
+    for resize_node, resize_node_idx in reversed(resize_nodes_to_insert_at_end):
+        usage_idx_list = node_to_usage_map[resize_node.args[0]]
+        pos_in_usage_idx_list = usage_idx_list.index(resize_node_idx)
+        for i in range(pos_in_usage_idx_list + 1, len(usage_idx_list)):
+            node = node_list[usage_idx_list[i]]
+            if node.target is torch.ops.aten.set_.source_Tensor and (
+                node.args[0] == resize_node.args[0]
+                or node.args[1] == resize_node.args[0]
+            ):
+                continue
+            if (
+                node.target is torch.ops.inductor.resize_storage_bytes_.default
+                and node.args[0] == resize_node.args[0]
+                and node.args[1] == 0
+            ):
+                continue
+            if node.target == "output":
+                continue
+            raise RuntimeError(
+                f"Input to .resize_storage_bytes_(X, 0) is used after the resize-to-0 op by `{node}`. "
+                f"Violating graph: {graph}"
+            )
+        if pos_in_usage_idx_list == len(usage_idx_list) - 1:
+            # If `.resize_(X, 0)` node is the last usage of X, we move the resize node to the end of the graph.
+            next_valid_usage_of_resized_tensor = return_node
+        else:
+            next_valid_usage_of_resized_tensor = node_list[
```

Comment:
Can you add a couple comments about what each step of the pass is doing? I think I'm interpreting this for loop as something like 

for each `resize_(x, 0)`, we will either:
(1) move it to the end of the graph if is the last use of `x`
(2) otherwise, `x` must have been used by either a `set_` or another `resize_` node late rin the graph.  If so, we move it right before its next usage.

Does that sound right? Also, doesn't that mean that we might still have some `resize_()` nodes in the middle of the graph after this pass has finished?

Reply:
removed all of this complexity, and we will no longer leave any `resize_()` in the middle of the graph :)